 specification/proto-requirements.md

```diff
@@ -0,0 +1,93 @@
+
+Examples of a Client are instrumented applications or sending side of telemetry collectors, examples of Servers are telemetry backends or receiving side of telemetry collectors (so a Collector is typically both a Client and a Server depending on which side you look from).
+
+## Known Issues with Existing Protocols
+
+Our experience with OpenCensus and other protocols has been that many of them have one or more of the following drawbacks:
+
+- High CPU consumption for serialization and especially deserialization of received telemetry data.
+- High and frequent CPU consumption by Garbage Collector.
+- Lack of delivery guarantees for certain protocols (e.g. stream-based gRPC OpenCensus protocol) which makes troubleshooting of telemetry pipelines difficult.
+- Not aware / not cooperating with load balancers resulting in potentially large imbalances in horizontally scaled backends.
+- Support either traces or metrics but not both.
+
+Our goal is to avoid or mitigate these known issues in the new protocol.
+
+## Requirements
+
+The following are OpenTelemetry protocol requirements.
+
+#### Supported Node Types
+
+The protocol must be suitable for use between all of the following node types: instrumented applications, telemetry backends, telemetry agents running as local daemons, stand-alone collector/forwarder services.
+
+#### Supported Data Types
+
+The protocol must support traces and metrics as data types.
+
+#### Reliability of Delivery
+
+The protocol must ensure reliable data delivery and clear visibility when the data cannot be delivered. This should be achieved by sending data acknowledgements from the Server to the Client. 
+
+Note that acknowledgements alone are not sufficient to guarantee that: a) no data will be lost and b) no data will be duplicated. Acknowledgements can help to guarantee a) but not b). Guaranteeing both at the same is difficult. Because it is usually preferable for telemetry data to be duplicated than to lose it, we choose to guarantee that there are no data losses while potentially allowing duplicate data.
+
+Duplicates can typically happen in edge cases (e.g. on reconnections, network interruptions, etc) when the client has no way of knowing if last sent data was delivered. In these cases the client will usually choose to re-send the data to guarantee the delivery which in turn may result in duplicate data on the server side.
+
+_To avoid having duplicates the client and the server could each track sent and delivered items using uniquely identifying ids. For this approach to be possible the client should be able to request the id of the last received data from the server and be able to resume sending from that item. Such request may be impossible to support on the server efficiently as it requires the server to have a consistent view of its data which may not be possible to do in a partitioned state, or it requires sticky sessions/connections which can be detrimental to load balancing and availability of servers._
```

Comment:
I agree with `To avoid having duplicates the client and the server could each track sent and delivered items using uniquely identifying ids`. I disagree with `For this approach to be possible the client should be able to request the id of the last received data from the server and be able to resume sending from that item`, which is not a must-have requirement.
De-dupe could happen at different places in the telemetry pipeline, either at agent, ingestion service, or the final storage. The client could keep sending the data (using a reasonable retry policy) until it received a success response.

Reply:
>De-dupe could happen at different places in the telemetry pipeline, either at agent, ingestion service, or the final storage. The client could keep sending the data (using a reasonable retry policy) until it received a success response.

I think what you say does not contradict to what this proposal says. The client will send the data again if it is not sure it was received. If the server wants to dedupe then that's server's responsibility, the client is not going to do anything special about it. 

The deduping you refer to that can happen at different places in the pipeline is orthogonal to the protocol and is not a protocol concern if protocol is not going to do anything about it.

Trace and Span Ids are already part of telemetry data. We don't need to add anything in the protocol to make deduplication possible. One can say that Ids are _already_ tracked, it is just not tracked at the protocol level, so from protocol's perspective we really do not care about the ids and are not going to do anything about ids or about preventing duplicates.

Reply:
I think we need two changes to the proposal:
1. Loose the statement `For this approach to be possible the client should be able to request the id of the last received data from the server and be able to resume sending from that item.` since it is one option but not a must-have thing.
2. Need some de-dupe id (for traces we can just reuse trace-id/span-id pair, for metrics/logs we need something else).

Reply:
OK, I'll remove 1 and will add a sentence explicitly mentioning that id tracking and deduping can be done at the layer above protocol layer.

Reply:
Fixed.